feat(#9869): add storageFree and storageTotal to user-devices API

najuna-brian opened the pull request:
<!--
Please use semantic PR titles that respect this format:
<type>(#<issue number>): <subject>
Quick example:
feat(#1234): add hat wobble
^--^(#^--^): ^------------^
|     |      |
|     |      + - > subject
|     |
|     + -------- > issue number
|
+ -------------- > type: chore, feat, fix, perf.
https://docs.communityhealthtoolkit.org/contribute/code/workflow/#commit-message-format
-->

# Description

Closes #9869

## Summary

This pull request enhances the `/api/v2/export/user-devices` endpoint by adding `storageFree` and `storageTotal` fields for each user device. This addresses the need for administrators to easily monitor device disk space, which can help in diagnosing performance issues related to low storage on Community Health Worker (CHW) devices.

Previously, obtaining this information was a hard process involving manual checks of telemetry documents in CouchDB or setting up complex data synchronization pipelines.

## Implementation Details

The following changes were made to implement this feature:

1. **CouchDB View Updated** (`ddocs/users-meta-db/users-meta/views/device_by_user/map.js`):
   * The map function was modified to extract `doc.device.deviceInfo.storage.free` and `doc.device.deviceInfo.storage.total` from telemetry documents.
   * These values are now emitted as part of the device information.

2. **API Service Modified** (`api/src/services/export/user-devices.js`):
   * The service now retrieves the `storage.free` and `storage.total` values from the data emitted by the CouchDB view.
   * These are included in the final JSON response as `storageFree` and `storageTotal` respectively.

3. **Tests Updated:**
   * **Unit Tests** (`api/tests/mocha/services/export-data.spec.js`): Updated to include mock data and assertions for the new `storageFree` and `storageTotal` fields in the `user-devices` export.
   * **Integration Tests** (`tests/integration/api/controllers/export-data.spec.js`):
      * The `createTelemetryDoc` helper was updated to allow inclusion of storage information.
      * Tests that interact with the `/api/v2/export/user-devices` endpoint were updated to verify the presence and correctness of the new storage fields in the API response.

## How to Test

Please follow the detailed testing steps outlined in issue #9869. Key verification points include:

1. Setting up a development environment and creating a test user with telemetry data containing storage information (as per the sample JSON in the issue).
2. Making a `curl` call to `http://medic:password@localhost:5988/api/v2/export/user-devices`.
3. Verifying that the output for the test user's device now includes the `storageFree` and `storageTotal` fields with the correct values from the telemetry document.

## Affected Files

* `api/src/services/export/user-devices.js`
* `api/tests/mocha/services/export-data.spec.js`
* `ddocs/users-meta-db/users-meta/views/device_by_user/map.js`
* `tests/integration/api/controllers/export-data.spec.js`

# Code review checklist
<!-- Remove or comment out any items that do not apply to this PR; in the remaining boxes, replace the [ ] with [x]. -->
- [ ] UI/UX backwards compatible: Test it works for the new design (enabled by default). And test it works in the old design, enable `can_view_old_navigation` permission to see the old design. Test it has appropriate design for RTL languages. 
- [ ] Readable: Concise, well named, follows the [[style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/)](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [[cht-docs](https://github.com/medic/cht-docs/)](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.
<!-- COMPOSE URLS GO HERE - DO NOT CHANGE -->

# License
The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

dianabarsan commented on api/src/services/export/user-devices.js:
```diff
@@ -18,19 +18,19 @@ module.exports = async () => {
-        name: browser?.name || undefined,
-        version: browser?.version || undefined,
-      },
+      browser: browser && browser.name && browser.version ? { name: browser.name, version: browser.version } : {},
```
Not sure why this change is necessary. There is even a test that checks that there is a browser property when these values are undefined.

dianabarsan commented on ddocs/users-meta-db/users-meta/views/device_by_user/map.js:
```diff
@@ -27,7 +27,34 @@ function(doc) {
+            doc.device.deviceInfo &&
+            doc.device.deviceInfo.storage &&
+var deviceInfo = doc.device && doc.device.deviceInfo;
+ emit([doc.metadata.user, doc.metadata.deviceId], {
```
Here you are emitting twice. I think you just added the code snippet I linked instead of incorporating it in the existent code.

dianabarsan commented on tests/integration/api/controllers/export-data.spec.js:
```diff
@@ -435,20 +449,12 @@ describe('Export Data V2.0', () => {
           versions: {
-            app: cht,
-            settings
+            app: cht, settings
```
This indentation change makes this harder to follow. Can you please revert it?

najuna-brian replied:
I have please reverted it to follow a cnsistent indentation

dianabarsan reviewed (commented):
It looks like your tests are not passing, because you have a linting failure. To get a faster turnaround, please run `npm run lint` before every push :)

najuna-brian commented:
@dianabarsan Thanks for the review! I've addressed all your feedback:

**CouchDB View Optimization**: Introduced `deviceInfo` variable to reduce null checks and improve readability as suggested
**Browser Property Logic**: Reverted the browser property changes to maintain existing behavior and compatibility  
**Code Integration**: Fixed the duplicate emit issue and properly integrated the storage fields into the existing emit block
**Indentation**: Reverted indentation changes to maintain consistency

All tests are passing 

Could you please re-review 
Thanks!

mrjones-plip commented:
@najuna-brian - I've kicked off a re-run of failed jobs in case  they  are indeed flaky

najuna-brian commented:
Thanks for retriggering those! @mrjones-plip

najuna-brian commented:
@mrjones-plip The retrigger shows `ci-webdriver-default-core` now passes (was failing)

mrjones-plip commented:
Hitting it again in hopes of the remaining 3 passing...

najuna-brian commented:
For the last 2 failing tests,

Fixed failing test: createTelemetryDoc was always creating storage objects, causing assertion mismatch (undefined vs [undefined] in view emissions).

Now conditionally creates storage only when values exist - aligns with real device telemetry patterns.

mrjones-plip commented:
re-running :crossed_fingers: !

najuna-brian commented:
Hi @mrjones-plip, @dianabarsan 
I would appreciate guidance on the newly pushed changes
Thank you

mrjones-plip commented:
I've merged `master` to this branch and I'll push the tests along if they get stuck, but otherwise letting the  natural review process do its thing.
